File: scripts/patches/patch-network-and-sync-fix.js
const fs = require('fs');
const path = require('path');

let logger;
try {
  logger = require('../utils/logger');
} catch (_) {
  logger = {
    info: (...args) => console.log('[INFO]', ...args),
    warn: (...args) => console.warn('[WARN]', ...args),
    error: (...args) => console.error('[ERROR]', ...args),
    success: (...args) => console.log('[SUCCESS]', ...args),
    dim: (...args) => console.log(' ', ...args)
  };
}

const APP_DIR = path.join(__dirname, '..', '..', 'app');

/**
 * Patch Zalo network state detection (gwig) and sync controller routing:
 * 1. Bypass CORS XMLHttpRequest failure to google.com in gwig network check,
 *    forcing getStateNetwork() to return CONNECTED so sync and calls proceed.
 * 2. Remove premature NO_NETWORK (1106) throw in main-startup.
 * 3. Route Sync Messages to SyncMessageController (V1) which sends push confirmation
 *    to mobile devices and saves message data via db-cross-v4 / sqlite3.
 * 4. Enable SQLite backup data adapter flag in main process.
 */
async function main() {
  const pcDistDir = path.join(APP_DIR, 'pc-dist');
  const mainDistDir = path.join(APP_DIR, 'main-dist');
  const lazyDir = path.join(pcDistDir, 'lazy');

  if (!fs.existsSync(pcDistDir)) {
    logger.warn('pc-dist directory not found, skipping network-and-sync fix');
    return;
  }

  function patchFile(filePath, replacements) {
    if (!fs.existsSync(filePath)) {
      return false;
    }

    let content = fs.readFileSync(filePath, 'utf8');
    let changed = false;

    for (const [from, to] of replacements) {
      if (typeof from === 'string') {
        if (content.includes(from)) {
          content = content.split(from).join(to);
          changed = true;
        }
      } else if (from instanceof RegExp) {
        if (from.test(content)) {
          content = content.replace(from, to);
          changed = true;
        }
      }
    }

    if (changed) {
      fs.writeFileSync(filePath, content, 'utf8');
      logger.dim(`Applied network & sync fix to ${path.basename(filePath)}`);
      return true;
    }
    return false;
  }

  function findFiles(dir, pattern) {
    if (!fs.existsSync(dir)) return [];
    return fs.readdirSync(dir)
      .filter(f => pattern.test(f))
      .map(f => path.join(dir, f));
  }

  // 1. main-startup bundle in lazy/
  const startupFiles = findFiles(lazyDir, /^main-startup\..*\.js$/);
  for (const f of startupFiles) {
    patchFile(f, [
      ['ge.b.getStateNetwork()===ge.a.DISCONNECT||_g.default.getSocketState()!==mg.l.OPEN', '!1'],
      [
        'checkFeatureEnabled(){return this.configService.isFeatureEnabled()?{ok:!0}:(this.logger.zsymb(9,"Oh4jaU",["Guard feature disabled by server config","VpsUI8"]),{ok:!1,reason:"feature_disabled"})}',
        'checkFeatureEnabled(){return{ok:!0}}'
      ],
      [
        'checkNetwork(){const e=ge.b.getStateNetwork();return e!==ge.a.DISCONNECT?{ok:!0}:(this.logger.zsymb(9,"3m6V8S",["Guard network disconnected","zonrE1"],{networkState:e}),{ok:!1,reason:"network_disconnected",networkState:e})}',
        'checkNetwork(){return{ok:!0}}'
      ],
      [
        'isEnable(){const e=this.config.get("cross_setting.offFeature"),t=this.config.get("cross_setting.enable");return!e&&t}',
        'isEnable(){return !0}'
      ],
      [
        'isEnableResume(){return!!this.isEnable()&&this.config.get("cross_setting.enableResume")}',
        'isEnableResume(){return !0}'
      ]
    ]);
  }

  // 2. default-login bundle in lazy/
  const defaultLoginFiles = findFiles(lazyDir, /^default-login-main-startup-shared-worker-znotification\..*\.js$/);
  for (const f of defaultLoginFiles) {
    patchFile(f, [
      ['const a=!0,s=!0,r=!0', 'const a=!0,s=!1,r=!0'],
      ['getStateNetwork(){return this.stateCur}', 'getStateNetwork(){return u.CONNECTED}'],
      ['_pingToDomain(e){return this._pingToDomainPC(e)}', '_pingToDomain(e){return Promise.resolve()}'],
      ['this.stateCur=u.NOT_SET', 'this.stateCur=u.CONNECTED'],
      ['canUseIpcCall(){return A.default.enable_ipc_call&&ne}', 'canUseIpcCall(){return !0}'],
      ['isSupport(){return!!A.default.enable_mac_call&&(A.default.enableCall&&se)}', 'isSupport(){return !0}'],
      ['isSupportVideoCall(){return this.isSupport()&&A.default.enableVideoCall}', 'isSupportVideoCall(){return !0}']
    ]);
  }

  // 3. Other worker & renderer bundles
  const otherBundles = [
    ...findFiles(pcDistDir, /^compact-app-pc\..*\.js$/),
    ...findFiles(pcDistDir, /^search-worker\..*\.js$/),
    ...findFiles(pcDistDir, /^sync-v2-sub-worker\..*\.js$/)
  ];
  for (const f of otherBundles) {
    patchFile(f, [
      ['const a=!0,i=!0,o=!0', 'const a=!0,i=!1,o=!0'],
      ['getStateNetwork(){return this.stateCur}', 'getStateNetwork(){return u.CONNECTED}'],
      ['_pingToDomain(e){return this._pingToDomainPC(e)}', '_pingToDomain(e){return Promise.resolve()}'],
      ['this.stateCur=u.NOT_SET', 'this.stateCur=u.CONNECTED'],
      ['canUseIpcCall(){return A.default.enable_ipc_call&&ne}', 'canUseIpcCall(){return !0}'],
      ['isSupport(){return!!A.default.enable_mac_call&&(A.default.enableCall&&ie)}', 'isSupport(){return !0}'],
      ['isSupportVideoCall(){return this.isSupport()&&A.default.enableVideoCall}', 'isSupportVideoCall(){return !0}']
    ]);
  }

  // 4. preload-sqlite.js
  const preloadSqlite = path.join(mainDistDir, 'preload-sqlite.js');
  if (fs.existsSync(preloadSqlite)) {
    patchFile(preloadSqlite, [
      ['cross_setting:{offFeature:!1,isMobileSupport:!1,', 'cross_setting:{enable:!0,offFeature:!1,isMobileSupport:!0,'],
      ['try{let t=1==e.settings.chat.enable_call;at.enableCall=t}catch(ht){}', 'try{let t=1==e.settings.chat.enable_call;at.enableCall=!0}catch(ht){}'],
      ['try{let t=1==e.settings.chat.enable_video_call;at.enableVideoCall=t}catch(ht){}', 'try{let t=1==e.settings.chat.enable_video_call;at.enableVideoCall=!0}catch(ht){}'],
      ['enable_group_call_for_user:0,enable_group_call_entry_for_group:0', 'enable_group_call_for_user:1,enable_group_call_entry_for_group:1']
    ]);
  }

  // 5. Enable SQLite backup data adapter flag in main process
  const mainJs = path.join(mainDistDir, 'main.js');
  if (fs.existsSync(mainJs)) {
    patchFile(mainJs, [
      [
        'const a=!1,s="--backupdata",c="ABORT_BACKING_UP_FOR_ALL_SESSION",l="CREATE_BACKING_UP_FOR_NO_SESSION";const d=!1',
        'const a=!0,s="--backupdata",c="ABORT_BACKING_UP_FOR_ALL_SESSION",l="CREATE_BACKING_UP_FOR_NO_SESSION";const d=!0'
      ]
    ]);
  }

  const utilSqlite = path.join(mainDistDir, 'utility-process-sqlite.js');
  if (fs.existsSync(utilSqlite)) {
    patchFile(utilSqlite, [
      [
        'const s=!1,a="--backupdata",c="ABORT_BACKING_UP_FOR_ALL_SESSION",u="CREATE_BACKING_UP_FOR_NO_SESSION";const l=!1',
        'const s=!0,a="--backupdata",c="ABORT_BACKING_UP_FOR_ALL_SESSION",u="CREATE_BACKING_UP_FOR_NO_SESSION";const l=!0'
      ]
    ]);
  }

  logger.success('Network and sync patches applied');
}

if (require.main === module) {
  main();
}

module.exports = { main };

File: scripts/patches/patch-network-and-sync-fix.py
#!/usr/bin/env python3
"""
Patch Zalo to fix network state detection (gwig) and sync controller routing.
1. Fix gwig network state detection: CORS XMLHttpRequests to google.com cause getStateNetwork() to return DISCONNECT (0).
   Force getStateNetwork() to return u.CONNECTED, initialize stateCur to u.CONNECTED, and make _pingToDomain always resolve.
2. Route "Sync Messages" to SyncMessageController (V1) which sends push confirmation to mobile and uses db-cross-v4.
3. Remove premature NO_NETWORK throw (1106) in main-startup.
"""

import os
import sys
import subprocess

def patch_file(file_path, replacements):
    print(f"[*] Patching {file_path}...")
    if not os.path.exists(file_path):
        print(f"[-] File not found: {file_path}")
        return False

    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    orig_content = content
    for pattern, replacement in replacements:
        count = content.count(pattern)
        if count == 0:
            print(f"    [WARN] Pattern not found: {pattern[:60]}...")
        else:
            content = content.replace(pattern, replacement)
            print(f"    [OK] Replaced ({count}x): {pattern[:50]}...")

    if content == orig_content:
        print(f"    [INFO] No changes made to {file_path}")
        return True

    with open(file_path, "w", encoding="utf-8") as f:
        f.write(content)

    print(f"[+] Saved changes to {file_path}")
    return True

def main():
    base_dir = "/home/truongit/zalo-build/app"

    # 1. main-startup: bypass premature NO_NETWORK (1106) throw
    main_startup = os.path.join(base_dir, "pc-dist/lazy/main-startup.c2df40f39fa8f98b69a8.js")
    patch_file(main_startup, [
        ("ge.b.getStateNetwork()===ge.a.DISCONNECT||_g.default.getSocketState()!==mg.l.OPEN", "!1")
    ])

    # 2. UI bundles: fix gwig network state and route sync to SyncMessageController (V1)
    default_login = os.path.join(base_dir, "pc-dist/lazy/default-login-main-startup-shared-worker-znotification.2b8b4ccdfaf363018e90.js")
    patch_file(default_login, [
        ("const a=!0,s=!0,r=!0", "const a=!0,s=!1,r=!0"),
        ("getStateNetwork(){return this.stateCur}", "getStateNetwork(){return u.CONNECTED}"),
        ("_pingToDomain(e){return this._pingToDomainPC(e)}", "_pingToDomain(e){return Promise.resolve()}"),
        ("this.stateCur=u.NOT_SET", "this.stateCur=u.CONNECTED")
    ])

    other_bundles = [
        os.path.join(base_dir, "pc-dist/compact-app-pc.0d138e05abbfee05885d.js"),
        os.path.join(base_dir, "pc-dist/search-worker.0d138e05abbfee05885d.js"),
        os.path.join(base_dir, "pc-dist/sync-v2-sub-worker.0d138e05abbfee05885d.js")
    ]

    for ui_file in other_bundles:
        patch_file(ui_file, [
            ("const a=!0,i=!0,o=!0", "const a=!0,i=!1,o=!0"),
            ("getStateNetwork(){return this.stateCur}", "getStateNetwork(){return u.CONNECTED}"),
            ("_pingToDomain(e){return this._pingToDomainPC(e)}", "_pingToDomain(e){return Promise.resolve()}"),
            ("this.stateCur=u.NOT_SET", "this.stateCur=u.CONNECTED")
        ])

    print("\n[*] Validating JS syntax with node --check...")
    all_files = [main_startup, default_login] + other_bundles
    for f in all_files:
        res = subprocess.run(["node", "--check", f], capture_output=True, text=True)
        if res.returncode == 0:
            print(f"  [PASS] {os.path.basename(f)}")
        else:
            print(f"  [FAIL] {os.path.basename(f)}")
            print("STDERR:")
            print(res.stderr[:500])
            sys.exit(1)

    print("\n[+] All patches applied and verified!")

if __name__ == "__main__":
    main()
